hcl/eval/partial.go

```diff
@@ -0,0 +1,1388 @@
+		}
+
+		e.emitnl()
+	}
+
+	if !e.hasTokens() {
+		panic("malformed object")
+	}
+
+	tok = e.peek()
+	if tok.Type != hclsyntax.TokenCBrace {
+		panic(errorf("malformed object, unexpected %s", tok.Bytes))
+	}
+
+	e.emit()
+	return nil
+}
+
+func (e *engine) evalForExpr(matchOpenType, matchCloseType hclsyntax.TokenType) error {
+	_, thisNode := e.newnode()
+	// { | [
+	tok := e.peek()
+	if tok.Type != matchOpenType {
+		panic(sprintf("evalForExpr: malformed `for` expression: %s", tok.Bytes))
+	}
+
+	e.emit()
+	e.emitnl()
+
+	// { for
+	tok = e.peek()
+	if tok.Type != hclsyntax.TokenIdent || string(tok.Bytes) != "for" {
+		panic(sprintf("evalForExpr: malformed `for` expression: %s", tok.Bytes))
+	}
+
+	e.emit()
+	// { for <ident>,<ident>,...
+	for e.hasTokens() && string(e.peek().Bytes) != "in" {
+		e.emitnl()
+		tok = e.peek()
+		if tok.Type != hclsyntax.TokenIdent {
+			return errorf("invalid `for` expression: wanted identifier, found %s", tok.Type)
+		}
+
+		e.emit()
+		e.emitnl()
+		tok = e.peek()
+		if tok.Type == hclsyntax.TokenComma {
+			e.emit()
+		}
+	}
+
+	tok = e.peek()
+	if tok.Type != hclsyntax.TokenIdent {
+		panic(errorf("found the `in` bytes of %s type instead of IDENT", tok.Type))
+	}
+
+	e.emit()
+
+	// consume everything and give errors in case of terramate variables being
+	// used in the `for`.
+	matchingCollectionTokens := 1
+	for e.hasTokens() && matchingCollectionTokens > 0 {
+		tok = e.peek()
+		switch tok.Type {
+		case matchOpenType:
+			matchingCollectionTokens++
+		case matchCloseType:
+			matchingCollectionTokens--
+		case hclsyntax.TokenQuestion:
+			thisNode.hasCond = true
+		}
+		v, found := e.parseVariable(e.tokens[e.pos:])
+		if found {
+			if v.isTerramate {
+				return errutil.Chain(
+					ErrForExprDisallowEval,
+					errorf("evaluating expression: %s", v.alltokens().Bytes()),
+				)
+			}
+
+			e.emitVariable(v)
+		} else {
+			e.emit()
+		}
+	}
+
+	return nil
+}
+
+func (e *engine) evalTmFuncall() error {
+	e.newnode()
+	if len(e.tokens[e.pos:]) < 3 {
+		return errorf("not a funcall")
+	}
+
+	begin := e.pos
+	tok := e.peek()
+	if !isTmFuncall(tok) {
+		panic("not a `tm_` function")
+	}
+
+	pos := e.skip(1)
+	if e.peekn(pos).Type != hclsyntax.TokenOParen {
+		panic(errorf("not a funcall: %s", e.tokens[pos:].Bytes()))
+	}
+
+	matchingParens := 1
+	e.pos += pos + 1
+	for e.hasTokens() {
+		switch e.peek().Type {
+		case hclsyntax.TokenOParen:
+			matchingParens++
+		case hclsyntax.TokenCParen:
+			matchingParens--
+		}
+
+		if matchingParens == 0 {
+			break
+		}
+
+		e.pos++
+	}
+
+	if matchingParens > 0 || e.peek().Type != hclsyntax.TokenCParen {
+		panic(errorf("malformed funcall: %s", e.tokens.Bytes()))
+	}
+
+	e.pos++
+
+	var expr []byte
+
+	for _, et := range e.tokens[begin:e.pos] {
+		expr = append(expr, et.Bytes...)
+	}
+
+	exprParsed, diags := hclsyntax.ParseExpression(expr, "gen.hcl", hcl.Pos{})
+	if diags.HasErrors() {
+		return errorf("failed to parse expr ('%s'): %v", expr, diags.Error())
```

Comment:
```suggestion
		panic(errorf("failed to parse expr ('%s'): %v", expr, diags.Error()))
```

Rationale: Isn't it an internal error/bug if after we checked the tokens and created the []byte we passed something invalid to ParseExpression ? Code was already initially parsed until we got here, so if something invalid is fed on ParseExpression it is probably a bug (terramate fault) and not that the input data was invalid (then it would return an error).